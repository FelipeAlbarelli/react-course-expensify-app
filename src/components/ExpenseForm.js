import React from 'react'
import moment from 'moment';
import {SingleDatePicker} from 'react-dates'


class ExpenseForm extends React.Component {
    
    constructor(props){
        super(props)
        
        this.state = {
            description : props.expense ? props.expense.description : '',
            amount : props.expense ? (props.expense.amount / 100).toString() : '',
            note : props.expense ? props.expense.note :  '',
            createdAt : props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused : false,
            error : ''
        }
    }
    
    state = {
    }

    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }));
      };

    onDateChange = createdAt => {
        // console.log(createdAt)
        this.setState( state => ({
            createdAt
        }) )
    }

    onDescriptionChange = e => {
        const description = e.target.value
        this.setState( state => ({
            description
        }) )
    }

    onNoteAreaChange = e => {
        const note = e.target.value;
        this.setState( state => ({
            note
        }) )
    }

    onAmountChange = e => {
        const amount = e.target.value;
        if (amount.match(/^\d*(\.\d{0,2})?$/)){
            this.setState( state => ({
                amount
            }) )
        }
    }

    onSubmitChange = e => {
        e.preventDefault();
        if (!this.state.description || !this.state.amount){
            this.setState( () => ({error : 'Provide description and amount'}) )
        } else {
            this.setState( () => ({error : ""}) )
            this.props.onSubmit({
                description : this.state.description,
                amount : parseFloat(this.state.amount)*100,
                note : this.state.note,
                createdAt : this.state.createdAt.valueOf()
            })
        }
    }

    render(){
        // console.log(expense)
        return (
            <div>
                {this.state.error}
                <form onSubmit={this.onSubmitChange} >
                    <input 
                        type='text' 
                        placeholder="Description" 
                        autoFocus 
                        value={this.state.description} 
                        onChange={this.onDescriptionChange} 
                    />
                    <input 
                        type='text' 
                        placeholder='Amount'
                        onChange={this.onAmountChange}
                        value={this.state.amount}
                    />
                    <SingleDatePicker 
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />
                    <textarea 
                        placeholder="Add a note for your expense"
                        value={this.state.note}
                        onChange={this.onNoteAreaChange}
                    ></textarea>
                    <button>Add Expense</button>
                </form>
            </div>
        )
    }
}

export default ExpenseForm