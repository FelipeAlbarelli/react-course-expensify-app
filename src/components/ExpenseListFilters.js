import React from 'react';
import {connect} from 'react-redux'
import {setTextFilter , sortByAmount , sortByDate} from '../actions/filters'

class ExpenseListFilters extends React.Component {

    constructor(props){
        super(props)
    }

    onTextChange = e => {
        this.props.setTextFilter(e.target.value)
    }

    onSortByChange = e => {
        const value = e.target.value;
        if (value == 'date'){
            this.props.sortByDate()
        } else {
            this.props.sortByAmount()
        }
    }

    render(){
        return (
            <div>
                <input type='text' value={this.props.filters.text} 
                onChange={this.onTextChange}
                />
                <select value={this.props.filters.sortBy} 
                onChange={this.onSortByChange} >
                    <option value='date' >Date</option>
                    <option value='amount' >Amount</option> 
                </select>
            </div>
        )
    }
}

const mapDispatchToProps = (dispacth) => ({
    setTextFilter : text => dispacth(setTextFilter(text)),
    sortByDate : () => dispacth(sortByDate()),
    sortByAmount : () => dispacth(sortByAmount())
})


const mapStateToProps = state => ({
    filters : state.filters
})

export {ExpenseListFilters};

export default connect(mapStateToProps ,mapDispatchToProps )(ExpenseListFilters)