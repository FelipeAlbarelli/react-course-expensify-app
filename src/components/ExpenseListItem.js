import React from 'react';
import moment from 'moment';
import {Link} from 'react-router-dom'

// description, amount, createdAt

const ExpenseListItem = ({ id ,description , amount, createdAt}) => (
    <div>
        <h3>{description}</h3>
        <p>{amount} - {moment(createdAt).format('DD/MM/YYYY')}</p>
        <Link to={`/edit/${id}`} ><button>edit</button></Link>
    </div>
)


export default ExpenseListItem