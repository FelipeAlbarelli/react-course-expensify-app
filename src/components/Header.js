import React from 'react';
import {NavLink} from 'react-router-dom'

const Header = () => (
    <header>
        <h1>Expensify</h1>
            <NavLink to='/' exact activeClassName='is-active'>Dashboard</NavLink>
            <NavLink to='/create' activeClassName='is-active'>Create</NavLink>
            <NavLink to='/help' activeClassName='is-active'>Help</NavLink>
            <NavLink to='/24234234' activeClassName='is-active'>404</NavLink>
    </header>
)

export default Header