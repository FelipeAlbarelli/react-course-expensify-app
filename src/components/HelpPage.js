import React from 'react'

const HelpPage = prop => (
    <div>
        Help Page
    </div>
)

export default HelpPage;