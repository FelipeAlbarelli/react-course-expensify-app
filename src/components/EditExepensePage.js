import React from 'react';
import {connect} from 'react-redux'
import ExpenseForm from './ExpenseForm'
import {editExpense , removeExpense} from '../actions/expenses'

export const EditExpensePage = props => {
    // console.log(props.expense)
    return (<div>
        <h1>Edit Expense</h1>
        <ExpenseForm
            expense={props.expense}
            onSubmit={ (expense) => {
                props.editExpense({ id : props.expense.id , updates : expense});
                props.history.push('/')
            } }
        />
        <button
            onClick={ () => {
                props.removeExpense({id : props.expense.id})
                props.history.push('/')
            } }
        >Remove</button>
    </div>)
}

const mapDispatchToProps = dispatch => ({
    editExpense : expense => dispatch(editExpense(expense)),
    removeExpense : expense => dispatch(removeExpense(expense))
})

const mapStateToProps = (state , props) => ({
    expense : state.expenses.find( expense => expense.id == props.match.params.id)
})

export default connect( mapStateToProps , mapDispatchToProps )(EditExpensePage);