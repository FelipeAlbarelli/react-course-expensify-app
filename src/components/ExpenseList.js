import React from 'react';
import {connect} from 'react-redux'
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

const ExpenseList = (props) => {
    return (
        <div>
            <h1>Expense List</h1>
            {
                props.expenses.length == 0 ? (
                    <p>no expenses :(</p>
                ) : (
                <div>
                    {props.expenses.map(
                        expense => <ExpenseListItem key={expense.id} {...expense} />
                    )}
                </div>
                )
            }
        </div>
    )
}

const ConnectedExpenseList = connect( state => {
    return {
        expenses : selectExpenses(state.expenses , state.filters)
    }
})(ExpenseList);

export {ExpenseList};
export default ConnectedExpenseList