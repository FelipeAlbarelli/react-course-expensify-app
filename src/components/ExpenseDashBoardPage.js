import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';

const ExpenseDashboardPage = prop => (
    <div>
        <ExpenseListFilters />
        <ExpenseList/>
    </div>
)

export default ExpenseDashboardPage