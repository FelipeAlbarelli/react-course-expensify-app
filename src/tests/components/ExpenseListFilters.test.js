import React from 'react';
import {shallow} from 'enzyme';
import {ExpenseListFilters} from '../../components/ExpenseListFilters' 

let wrapper , setTextFilter , sortByDate , sortByAmount;

const filters = {
    text : "",
    sortBy : 'date'
}

const altFilters = {
    text : 'some text',
    sortBy : 'amount'
}

beforeEach( () => {
    sortByAmount = jest.fn();
    sortByDate = jest.fn();
    setTextFilter = jest.fn();
    wrapper = shallow(
        <ExpenseListFilters 
            filters={filters}
            sortByAmount={sortByAmount}
            sortByDate={sortByDate}
            setTextFilter={setTextFilter}
        />
    )
} )

test('should render ExpenseListFilters correctly' , () => {
    expect(wrapper).toMatchSnapshot()
})

test('should render ExpenseListFilters with altFilters correctly' , () =>{
    wrapper.setProps({
        filters : altFilters
    })
    expect(wrapper).toMatchSnapshot()
})

// text change , sortByAmount , sortByDate

test('should call setTextFilter on textChange' , ()=>{
    wrapper.find('input').simulate('change' , {target : { value : 'some str' }});
    expect(setTextFilter).toHaveBeenLastCalledWith('some str')
})

test('should call sortyAmount on selectChange' , () => {
    wrapper.find('select').simulate('change' , {target : {value : 'amount'}});
    expect(sortByAmount).toHaveBeenCalled()
})

test('should call sortyDate on selectChange' , () => {
    wrapper.find('select').simulate('change' , {target : {value : 'date'}})
    expect(sortByDate).toHaveBeenCalled()
})

