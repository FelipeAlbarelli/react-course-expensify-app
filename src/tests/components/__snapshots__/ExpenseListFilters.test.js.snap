// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseListFilters correctly 1`] = `
<div>
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
  <select
    onChange={[Function]}
    value="date"
  >
    <option
      value="date"
    >
      Date
    </option>
    <option
      value="amount"
    >
      Amount
    </option>
  </select>
</div>
`;

exports[`should render ExpenseListFilters with altFilters correctly 1`] = `
<div>
  <input
    onChange={[Function]}
    type="text"
    value="some text"
  />
  <select
    onChange={[Function]}
    value="amount"
  >
    <option
      value="date"
    >
      Date
    </option>
    <option
      value="amount"
    >
      Amount
    </option>
  </select>
</div>
`;
